### DISCUSSION TITLE
Gauss-Newton Hessian with Knitro

### DISCUSSION THREAD

**Discussion by richteropt** [casadi/casadi#3110] [General]
Hi!

Yet another question related to Knitro: Is there any way to provide Knitro with a user-defined Hessian, e.g. Gauss-Newton? The way via 'hess_lag' is not working for Knitro, only for Ipopt.

Thanks.

**Comment by jaeandersson** [comment]
In the 3.6.0 release, there is a new option in the `Function` base class called `cache`. It's more generic than the IPOPT specific option "hess_lag", and we'll probably remove the "hess_lag" approach for in a future release.

For Knitro, you need to set the "nlp_hess_l" entry of the `cache` option. The IO of your Gauss-Newton Hessian needs to match the existing one - since this is a low-level functionality there is very little testing.

Syntax should be something like this:
```python
# Hessian expression
x = ...
p = ...
lam_f = ...
lam_g = ...
H = ...
# Function that calculates the Hessian of the Lagrangian (upper triangular part only)
nlp_hess_l = Function('nlp_hess_l', [x, p, lam_f, lam_g], [triu(H)], ['x', 'p', 'lam_f', 'lam_g'], ['H'])
# Pass to solver
mysolver = casadi.nlpsol('mysolver', 'knitro', NLP, dict(cache = dict(nlp_hess_l = nlp_hess_l)))
```
Good luck.

**Reply by richteropt** [reply]
Thanks for your quick answer. I am working with Matlab and there I do it like

nlp_hess_l = Function('nlp_hess_l', {qq; [vec(R_tau); p_tau]; sigma; lam_g}, {sigma*triu(HH)}, {'x', 'p', 'lam_f', 'lam_g'}, {'H'});
options.cache.nlp_hess_l = nlp_hess_l;

However, if I run it, I obtain
CasADi - 2023-04-14 17:54:42 WARNING("Cannot add 'nlp_hess_l' a.k.a. 'nlp_hess_l' to cache. Mismatching names not implemented.") [.../casadi/core/function_internal.cpp:620]

Any idea what is happening there?

Thanks.

**Comment by jaeandersson** [comment]
Ugh, I think there is a bug in the code (it's really an internal feature at this point). I think you can work around the bug by adding a ":" at the end of the struct name:
```matlab
options.cache = struct('nlp_hess_l:', nlp_hess_l);
```

**Reply by richteropt** [reply]
Unfortunately, a struct fieldname is not allowed to contain a ':'...

**Reply by jaeandersson** [reply]
Interestingly, Octave behaves differently in this case it seems. Created an issue for fixing it here: https://github.com/casadi/casadi/issues/3111. Should hopefully make it in to 3.6.1. due in a few days.

**Reply by richteropt** [reply]
Thanks!
